R/utilities.R

```diff
@@ -19,18 +20,39 @@ set_axis_label <- function(p, xlab, ylab, p2 = NULL) {
     }
 
     if (has_theme) {
-        xlab_param <- p2$theme$axis.title.x
-        ylab_param <- p2$theme$axis.title.y
-        
-        p <- p + theme(axis.title.x = do.call(element_text, xlab_param),
-                       axis.title.y = do.call(element_text, ylab_param))
+        x <- p2
     } else {
-        p <- p + theme(axis.title.x = element_text(vjust = 1),
-                       axis.title.y = element_text(angle = 90, vjust = 1)) 
+        x <- NULL
     }
+    xlab_param <- get_theme_params(x=x, i="axis.title.x")
+    ylab_param <- get_theme_params(x=x, i="axis.title.y")
+    title_param <- get_theme_params(x=x, i="plot.title")
+    title_pos_param <- get_theme_params(x=x, i="plot.title.position")
+    subtitle_param <- get_theme_params(x=x, i="plot.subtitle")
+    cap_param <- get_theme_params(x=x, i="plot.caption")
+    cap_pos_param <- get_theme_params(x=x, i="plot.caption.position")
+    tag_param <- get_theme_params(x=x, i="plot.tag")
+    tag_pos_param <- get_theme_params(x=x, i="plot.tag.position")
+    p <- p + theme(axis.title.x = xlab_param,
+                   axis.title.y = ylab_param,
+                   plot.title = title_param,
+                   plot.title.position = title_pos_param,
+                   plot.subtitle = subtitle_param,
+                   plot.caption = cap_param,
+                   plot.caption.position = cap_pos_param,
+                   plot.tag = tag_param,
+                   plot.tag.position = tag_pos_param
+                  )
```

Comment:
something like this would be much better.
```
> xx
[1] "axis.title.x" "axis.title.y"
> aa=setNames(lapply(xx, get_theme_params, x=p), xx)
> p + do.call(theme, aa)
```

Reply:
```
get_theme_params = function(x, i, as_theme=TRUE) {
    if (!inherits(x, "theme")) x <- x$theme
    if (length(x) == 0) {
        x <- ggplot2::theme_get()
    }
    
    x[i]
}

theme_fp <- function(x, i) {
    params <- get_theme_params(x, i)
    do.call(theme, params)
}
```

even better.

then you can just use p + theme_fp(p2, aa). fp stands for format painter.